### DISCUSSION TITLE
Mechanism to query role hierarchy from security voters

### DISCUSSION THREAD

**Discussion by Bilge** [symfony/symfony#42889] [Q&A] [answered]
**Description**
A mechanism to query the role hierarchy from security voters is needed.

The [role hierarchy](https://symfony.com/doc/current/security.html#security-role-hierarchy) allows *implicit* roles to be set for other *explicit* roles. However, the only way to know if an implicit role is effective for the current user is by querying `Security::isGranted`. It should be expected that security voters may wish to know if a user has a particular role, whether implicit or explicit, in which case the only way to know this is by again querying `Security::isGranted`. However, `isGranted` invokes the voters, so this creates an infinite loop. Therefore, there appears to be no way for voters to know if a user has implicit roles granted by the role hierarchy.

**Example**
```yaml
security:
  role_hierarchy:
    ROLE_FOO:
      - ROLE_BAR

```

```php
final class MyVoter implements VoterInterface
{
    public function __construct(private Security $security) {}

    public function vote(TokenInterface $token, $subject, array $attributes): int
    {
        return $this->security->isGranted('ROLE_BAR'); // Infinite loop.
    }
}
```

*Technically, the infinite loop may not occur if the current user has the explicit role that grants the implicit role being queried, since role hierarchy voters appear to be registered earlier than custom voters and thus short circuit the loop in time. Querying for a role the current user does not have will definitely result in an infinite loop, though.*

**Comment by derrabus** [comment, marked as answer]
I have transferred this issue to support because there is already a documented way to do this: https://symfony.com/doc/current/security/voters.html#checking-for-roles-inside-a-voter

Querying the `Security` façade should work. This is a recursive call (if that's what you meant by "infinite loop"), but usually the recursion should terminate if all your voters play well.

The voter in your code example however is broken: it will perform the recursive call not matter what permission is checked. It should at least abstain on roles, if you want the recursion to terminate. Ideally, it should whitelist the permissions it performs a recursive check for.

**Reply by Bilge** [reply]
Thanks for your prompt reply. I see now. Of course, my example was entirely contrived, but I can nevertheless work within those constraints if that is how it is intended to be used. 

Another idea I had was to just declare everything within the role hierarchy configuration and avoid custom voters altogether. Whilst a nice idea, this is not possible unless all your attributes are prefixed with `ROLE_` since the `RoleHierarchyVoter` created by Symfony is hard-coded to only respond to this prefix and offers no way to change it via YAML configuration.

**Reply by derrabus** [reply]
Yeah, you shouldn't register all your permissions as roles, really.

Custom voters for custom permissions are totally fine. 🙂

**Reply by Bilge** [reply]
👍🏻
